### PR TITLE
feat: PRT - Optimizer metrics - including tiers

### PR DESCRIPTION
**Pull request by shleikes**

### CODE REVIEW

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: protocol/lavasession/consumer_session_manager.go [hunk 1]
```diff
@@ -1126,7 +1146,32 @@ func (csm *ConsumerSessionManager) GenerateReconnectCallback(consumerSessionsWit
...
 
-func NewConsumerSessionManager(rpcEndpoint *RPCEndpoint, providerOptimizer ProviderOptimizer, consumerMetricsManager *metrics.ConsumerMetricsManager, reporter metrics.Reporter, consumerPublicAddress string, activeSubscriptionProvidersStorage *ActiveSubscriptionProvidersStorage) *ConsumerSessionManager {
+func (csm *ConsumerSessionManager) periodicCollectOptimizerProvidersScore(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(CollectOptimizerProvidersScoreInterval):
+			// collect optimizer providers score
+			selectionTier, _ := csm.providerOptimizer.CalculateSelectionTiers(csm.validAddresses, nil, 10, spectypes.LATEST_BLOCK)
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in 66f074bd01485842fe48d8868985f25391163666

[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in eacc61bbe2f3edd6e2c67cdeda472711b7d93e48

**Comment by omerlavanet** [comment]
@shleikes if this is stale please remove the PR

File: protocol/lavasession/consumer_session_manager.go [hunk 2]
```diff
@@ -27,8 +27,12 @@ const (
 )
 
 var (
-	retrySecondChanceAfter = time.Minute * 3
-	DebugProbes            = false
+	retrySecondChanceAfter                     = time.Minute * 3
+	DebugProbes                                = false
+	CollectOptimizerProvidersScore             = false
+	CollectOptimizerProvidersScoreFlagName     = "collect-optimizer-providers-score"
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in 8bde3988f0b52cb9db9ee6c4537f30f23305e9b0

File: protocol/lavasession/consumer_session_manager.go [hunk 3]
```diff
@@ -114,6 +118,19 @@ func (csm *ConsumerSessionManager) UpdateAllProviders(epoch uint64, pairingList
 	csm.setValidAddressesToDefaultValue("", nil) // the starting point is that valid addresses are equal to pairing addresses.
 	// reset session related metrics
 	csm.consumerMetricsManager.ResetSessionRelatedMetrics()
+	stakeEntriesForMetrics := map[string]uint64{}
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in ee19d51f05192fdafaa745edbbda21cd7c09af60

File: protocol/lavasession/consumer_session_manager.go [hunk 4]
```diff
@@ -639,7 +656,10 @@ func (csm *ConsumerSessionManager) getValidProviderAddresses(ignoredProvidersLis
 	if stateful == common.CONSISTENCY_SELECT_ALL_PROVIDERS && csm.providerOptimizer.Strategy() != provideroptimizer.STRATEGY_COST {
 		providers = csm.getTopTenProvidersForStatefulCalls(validAddresses, ignoredProvidersList)
 	} else {
-		providers, _ = csm.providerOptimizer.ChooseProvider(validAddresses, ignoredProvidersList, cu, requestedBlock)
+		providers, _ = csm.providerOptimizer.ChooseProvider(validAddresses, ignoredProvidersList, cu, requestedBlock, csm.currentEpoch)
+		for _, chosenProvider := range providers {
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in b9d7e5dd81b0a36db54b3d68e185a385b8034e1e

File: protocol/lavasession/consumer_session_manager.go [hunk 5]
```diff
@@ -1068,7 +1088,7 @@ func (csm *ConsumerSessionManager) updateMetricsManager(consumerSession *SingleC
 	publicProviderAddress := consumerSession.Parent.PublicLavaAddress
 
 	go func() {
-		csm.consumerMetricsManager.SetQOSMetrics(chainId, apiInterface, publicProviderAddress, lastQos, lastQosExcellence, consumerSession.LatestBlock, consumerSession.RelayNum, relayLatency, sessionSuccessful)
+		csm.consumerMetricsManager.SetQOSMetrics(chainId, apiInterface, publicProviderAddress, lastQos, lastQosExcellence, consumerSession.LatestBlock, consumerSession.RelayNum, relayLatency, sessionSuccessful, csm.currentEpoch)
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in ab36352e952e9032cdc3bb9e7842b9ca1f3701b5

File: protocol/lavasession/consumer_session_manager.go [hunk 6]
```diff
@@ -1126,7 +1146,32 @@ func (csm *ConsumerSessionManager) GenerateReconnectCallback(consumerSessionsWit
...
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(CollectOptimizerProvidersScoreInterval):
+			// collect optimizer providers score
+			selectionTier, _ := csm.providerOptimizer.CalculateSelectionTiers(csm.validAddresses, nil, 10, spectypes.LATEST_BLOCK)
+			metricsTiers := []metrics.ProviderTierEntry{}
+			for i := 0; i < provideroptimizer.OptimizerNumTiers; i++ {
+				tierEntries := selectionTier.GetTier(i, provideroptimizer.OptimizerNumTiers, 1)
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in b95df52df43c635b5bb8401933d9f1247e4beb82

File: protocol/lavasession/consumer_session_manager.go [hunk 7]
```diff
@@ -1126,7 +1146,32 @@ func (csm *ConsumerSessionManager) GenerateReconnectCallback(consumerSessionsWit
...
 }
 
-func NewConsumerSessionManager(rpcEndpoint *RPCEndpoint, providerOptimizer ProviderOptimizer, consumerMetricsManager *metrics.ConsumerMetricsManager, reporter metrics.Reporter, consumerPublicAddress string, activeSubscriptionProvidersStorage *ActiveSubscriptionProvidersStorage) *ConsumerSessionManager {
+func (csm *ConsumerSessionManager) periodicCollectOptimizerProvidersScore(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(CollectOptimizerProvidersScoreInterval):
+			// collect optimizer providers score
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go:1173]: Fixed in 8a21e28d727afd904d259c944c28beb69653de2f

File: protocol/lavasession/consumer_session_manager.go [hunk 8]
```diff
@@ -1126,7 +1146,32 @@ func (csm *ConsumerSessionManager) GenerateReconnectCallback(consumerSessionsWit
...
+				for _, entry := range tierEntries {
+					metricsTiers = append(metricsTiers, metrics.ProviderTierEntry{
+						Address: entry.Address,
+						Score:   entry.Score,
+						Tier:    i,
+					})
+				}
+			}
+
+			go csm.consumerMetricsManager.UpdateOptimizerProvidersScore(csm.rpcEndpoint.ChainID, csm.rpcEndpoint.ApiInterface, csm.currentEpoch, metricsTiers)
```
[the remark this answers is not in the document]
Comment (shleikes) [reply protocol/lavasession/consumer_session_manager.go]: Fixed in 8a21e28d727afd904d259c944c28beb69653de2f

**Pull request closed without merging**